distributed/worker.py

```diff
@@ -3102,7 +3106,11 @@ async def gather_dep(
             refresh_who_has = set()
 
             for d in self.in_flight_workers.pop(worker):
-                ts = self.tasks[d]
+                try:
+                    ts = self.tasks[d]
+                except KeyError:
+                    continue
```

Comment:
Tasks are not allowed to be released while in flight. This is why I introduced the `cancelled` state. Releasing at this stage, i.e. forgetting can lead to all sorts of race conditions

Reply:
What do you make of the attached story?  It seems like they are being forgotten in between being placed in `in_flight_workers` and this stage (there is an await in there, during which I expect pretty much anything can happen).  It seems like they're getting cancelled, then resumed, then released and forgotten all in the time span created by the `await` above.

```
("('rechunk-split-82117560f6f829a7fa07bfef62cff7d5', 1006)", 'flight', 'released', 'cancelled', {}, 'processing-released-1650891186.0754924', 1650891186.16295)
("('rechunk-split-82117560f6f829a7fa07bfef62cff7d5', 1006)", 'compute-task', 'compute-task-1650891186.392905', 1650891186.4060445)
("('rechunk-split-82117560f6f829a7fa07bfef62cff7d5', 1006)", 'cancelled', 'waiting', 'cancelled', {"('rechunk-split-82117560f6f829a7fa07bfef62cff7d5', 1006)": ('resumed', 'waiting')}, 'compute-task-1650891186.392905', 1650891186.40608)
("('rechunk-split-82117560f6f829a7fa07bfef62cff7d5', 1006)", 'cancelled', 'resumed', 'resumed', {}, 'compute-task-1650891186.392905', 1650891186.406102)
('free-keys', ("('rechunk-split-82117560f6f829a7fa07bfef62cff7d5', 1006)",), 'processing-released-1650891188.465502', 1650891188.5917683)
("('rechunk-split-82117560f6f829a7fa07bfef62cff7d5', 1006)", 'release-key', 'processing-released-1650891188.465502', 1650891188.5917764)
("('rechunk-split-82117560f6f829a7fa07bfef62cff7d5', 1006)", 'resumed', 'released', 'released', {"('rechunk-split-82117560f6f829a7fa07bfef62cff7d5', 1006)": 'forgotten'}, 'processing-released-1650891188.465502', 1650891188.5917976)
("('rechunk-split-82117560f6f829a7fa07bfef62cff7d5', 1006)", 'released', 'forgotten', 'forgotten', {}, 'processing-released-1650891188.465502', 1650891188.591806)
```

Reply:
Sounds good.  As a heads-up I'm pausing work here to go play with https://github.com/dask/distributed/issues/6206 .  The extent to which you can take this over would be welcome.

The full story is in the linked issue if that provides any more insight.